mex.go

```diff
@@ -43,28 +46,85 @@ const (
 	mexChannelBufferSize = 2
 )
 
+type errNotifier struct {
+	c   chan struct{}
+	err error
+}
+
+func newErrNotifier() errNotifier {
+	return errNotifier{c: make(chan struct{})}
+}
+
+// Notify will store the error and notify all waiters on c that there's an error.
+func (e *errNotifier) Notify(err error) error {
+	// The code should never try to Notify(nil).
+	if err == nil {
+		panic("cannot Notify with no error")
+	}
+
+	// There may be some sort of race where we try to notify the mex twice.
+	select {
+	case <-e.c:
+		return fmt.Errorf("cannot broadcast error: %v, already have: %v", err, e.err)
+	default:
+	}
```

Comment:
It didn't protect against races very well, it was mostly to ensure we didn't call it twice. But better to do it properly, so updated to use an atomic int (it's less code, and more readable than locks IMO)